### DISCUSSION TITLE
v0.16.0: Dark mode and modifier functions parsing

### DISCUSSION THREAD

**Discussion by caksoylar** [caksoylar/keymap-drawer#92] [Announcements]
This release contains two major new features: parsing of modifier functions and dark mode support! It also contains a couple of smaller breaking changes, so make sure to check them out below.

### Parsing
- Added support for **parsing modifier functions** like `LC(V)` in ZMK and `LCTL(KC_V)` in QMK, used for sending modified keycodes like <kbd>ctrl</kbd>+<kbd>v</kbd> events with one key press
  - By default these will be converted to `Ctl+ V` representation, this is configurable via the new [`parse_config.modifier_fn_map` setting](https://github.com/caksoylar/keymap-drawer/blob/v0.16.0/CONFIGURATION.md#modifier_fn_map)
  - You can also assign special names to specific mod combinations, like `Hyper+ X` for `LC(LS(LG(LA(X))))`
  - **Breaking change:** If you were using entries like `LC(V): Paste` in `parse_config.zmk_keycode_map` or `C(V): Paste` in `qmk_keycode_map`, they will no longer work because `LC` and `V` parts are processed separately. If you want to preserve the mapping, instead define them in `raw_binding_map` like `&kp LC(V): Paste` for ZMK or `C(KC_V): Paste` for QMK.
- Added parsing for QMK tap-toggle layers `TT(LAYER)` by @m-demare in https://github.com/caksoylar/keymap-drawer/pull/87

### Drawing
- Added **dark mode support**, by @snoyer!
  - This is implemented as a special set of [CSS overrides](https://github.com/caksoylar/keymap-drawer/blob/v0.16.0/CONFIGURATION.md#svg_style_dark1) that get added to the final CSS if the new `draw_config.dark_mode` setting is on, or added under `@media (prefers-color-scheme: dark)` when set to `"auto"` to fit the web page or OS light/dark setting.
  - The default is `false` which means light mode only, so this is not a breaking change for CLI users. The [web app](https://caksoylar.github.io/keymap-drawer/) overrides it to `"auto"`.

| `dark_mode: false` | `dark_mode: true` | `dark_mode: auto` |
| - | - | - |
| ![logo_light](https://github.com/caksoylar/keymap-drawer/assets/7876996/3ed5e809-53d0-477f-b822-7ed66904eb53) | ![logo_dark](https://github.com/caksoylar/keymap-drawer/assets/7876996/40091ef2-af09-4165-9be6-cc958d9fc7b7) | ![logo](https://github.com/caksoylar/keymap-drawer/assets/7876996/83da01c4-690c-471d-bf7d-a5d7a2a428da) |

- Added support for **appending footer text** to produced SVGs using [`draw_config.footer_text`](https://github.com/caksoylar/keymap-drawer/blob/v0.16.0/CONFIGURATION.md#footer_text)
  - The value goes into a `<text>` element but it isn't escaped, so you can add e.g. links using `<a>` tags
  - The default value is empty so this isn't a breaking change for CLI users. The [web app](https://caksoylar.github.io/keymap-drawer/) overrides it to "<text>Created with <a href="https://github.com/caksoylar/keymap-drawer">keymap-drawer</a></text>"

- Added support for **using [Phosphor Icons](https://phosphoricons.com)**, check out how to use them in [the README section](https://github.com/caksoylar/keymap-drawer/blob/v0.16.0/README.md#custom-svg-glyphs)

### CLI

- Added `-o`/`--output` parameter to all `keymap` subcommands, to write their output to the given path rather than stdout
  - Using these parameters are recommended if you are running the CLI on Windows, since they ensure the outputs are written with `utf-8` encoding rather than the locale-specific one (which can break many things). This release also incorporates a couple fixes that ensure operation in Windows (tested in powershell)
  - **Breaking change:** `keymap draw --ortho-layout` no longer supports the `-o` short form since that is taken over by the output parameter

**Full Changelog**: https://github.com/caksoylar/keymap-drawer/compare/v0.15.0...v0.16.0

<hr /><em>This discussion was created from the release <a href='https://github.com/caksoylar/keymap-drawer/releases/tag/v0.16.0'>v0.16.0: Dark mode and modifier functions parsing</a>.</em>

**Comment by haunt98** [comment]
Thank you @caksoylar, such a wonderful update. I will try dark mode ASAP :D

**Comment by caksoylar** [comment]
It looks like Safari still doesn’t support SVGs inheriting color scheme, unfortunately: https://bugs.webkit.org/show_bug.cgi?id=199134

So auto mode always resolves to light variant on Safari.

**Reply by snoyer** [reply]
> It looks like Safari still doesn’t support SVGs inheriting color scheme, unfortunately: [bugs.webkit.org/show_bug.cgi?id=199134](https://bugs.webkit.org/show_bug.cgi?id=199134)

So Safari does supports `prefers-color-scheme` (https://caniuse.com/prefers-color-scheme) but not for SVG? :(

> So auto mode always resolves to light variant on Safari.

Technically there could be an alternative auto mode that does the overriding the other way (dark by default, changes to light if preferred and supported) using `@media (prefers-color-scheme: light)`. Unfortunately that doesn't solve the problem so I'm not sure it would be worth the added complexity.

**Reply by caksoylar** [reply]
Yeah, I am inclined to let this go. Maybe some tricks can be done to detect the color scheme in md/html and show one of the two SVGs when you want to embed them, but that's out of scope for the tool itself.
